### PR TITLE
[sw,otbn] rename OTBN scripts so they can be bazelified

### PR DESCRIPTION
**Pull request by timothytrippel**
Building the mask ROM with bazel requires building some OTBN code, which requires using OTBN specific tools, including the OTBN linker and assembler. These are implemented as Python scripts. However, to bazelify them, they must have names that end in a proper ".py" extension. This renames said tools so that bazel `py_binary` targets can be created for them, so bazel can use the hermetic python toolchain to run them properly.

### CODE REVIEW

**Review by rswarbrick — CHANGES_REQUESTED** [review]
I really dislike the changed names: the fact that we've implemented our assembler/linker/objdump wrappers in Python is supposed to be an implementation detail.

Please could you rejig this so that e.g. `otbn-as` still exists? (It could even be just a symlink to `otbn_as.py` if that's easiest).

**Comment by rswarbrick** [comment]
As a more general point, I don't think I really understand the motivation behind this. Is there really no way for Bazel to understand an executable Python script that doesn't happen to be importable? That sounds pretty bizarre to me...

**Review by mcy — APPROVED** [review]

**Comment by mcy** [comment]
Rupert: this is a hard Bazel limitation. Working around this will delay the Bazel migration further, and I don't think we should be removing file extensions for the purpose of aesthetics is a good choice anyways.

**Comment by rswarbrick** [comment]
It's impossible to represent a program that happens to be written in Python if it doesn't have a specific filename? I do not believe that can possibly be true.

**Comment by timothytrippel** [comment]
Unfortunately, when I create a `py_binary` with a src file that does not have a `.py` extension, like below:
Screen Shot 2022-04-21 at 7 54 42 AM

I get the following error:
Screen Shot 2022-04-21 at 7 54 02 AM

I too was a little surprised by this, that the bazel `py_binary` rule would attempt to enforce such constraints. When I went to track down the line of starlark that was causing this, I found that the `py_binary` rule is [really just using the `native.py_binary`](https://github.com/bazelbuild/rules_python/blob/4c7e63fe6842343f3fb30cd744aaba54596f537d/python/defs.bzl#L62), and I gave up searching for its implementation after some time. As the error suggested, I renamed the files and the error went away.

I could certainly check in a symlink connecting `hw/ip/otbn/util/otbn_as.py` to `hw/ip/otbn/util/otbn-as`, but unfortunately, it seems that in order to encode the `hw/ip/otbn/util/otbn-as` as a py_binary, which I need to do in order to build the mask ROM in an air-gapped system where bazel only knows about its hermetic python toolchain and python packages, then I have to rename `hw/ip/otbn/util/otbn-as` to `hw/ip/otbn/util/otbn_as.py`.

The reason I did not also check in a symlink was that it seemed unnecessary to store / maintain an additional pointer to the same file in a VCS. If that is an unpopular opinion though I am more than happy to add a symlink.

**Comment by cfrantz** [comment]
> As a more general point, I don't think I really understand the motivation behind this. Is there really no way for Bazel to understand an executable Python script that doesn't happen to be importable? That sounds pretty bizarre to me...

From bazel's point-of-view, these are not executable scripts.  They're source code to programs.  The outputs of building the programs are extensionless hermetic binaries (e.g. `bazel-bin/hw/ip/otbn/util/otbn_as`).  In fact, you can write rules which will bundle the resutling binary into a single executable archive of all needed resources (similary to py2exe or cxFreeze from days of yore).

All of the files in the source tree (//sw/..., //hw/..., etc) should be _source code_ and bazel should be taught how to build that code into executables.

File: doc/ug/otbn_sw.md [hunk 1]
```diff
@@ -23,19 +23,19 @@ For specific formatting and secure coding guidelines, see the [OTBN style guide]
 
 ### Assembler
 
-The OTBN assembler is called `otbn-as` and can be found at `hw/ip/otbn/util/otbn-as`.
+The OTBN assembler is called `otbn_as.py` and can be found at `hw/ip/otbn/util/otbn_as.py`.
```
Comment (cfrantz) [inline doc/ug/otbn_sw.md:26]: It could be argued that the assembler is now called `//hw/ip/otbn/util:otbn_as` (and similarly for the linker).

**Review by cfrantz — APPROVED** [review]

File: doc/ug/otbn_sw.md (hunk 1 above)
Comment (mcy) [reply doc/ug/otbn_sw.md:26]: That's what its name morally always was, only it was not hermetic. :)

Hopefully in the fullness of time we can eliminate almost all shebangs in the repo.

**Comment by mcy** [comment]
@timothytrippel btw, the reason for this is because py_binary specifies `allow_files` in an `attr.label_list` call that filters out non-python files: https://docs.bazel.build/versions/main/skylark/lib/attr.html#label_list. This is common convention in pretty much every "compile stuff" rule.

**Comment by rswarbrick** [comment]
For the record, my review still stands. Thanks, @timothytrippel for explaining the problem. I'd have appreciated the chance to reply to the PR before my review was dismissed and the PR was merged.

**Pull request merged**